stdlib/jdbc/src/main/java/org/ballerinax/jdbc/datasource/PoolKey.java

```diff
@@ -0,0 +1,111 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.ballerinax.jdbc.datasource;
+
+import org.ballerinalang.jvm.TypeChecker;
+import org.ballerinalang.jvm.types.BType;
+import org.ballerinalang.jvm.types.TypeTags;
+
+import java.util.Map;
+
+/**
+ * The key that uniquely identifies a connection pool.
+ */
+public class PoolKey {
+    private String jdbcUrl;
+    private Map<String, ?> dbOptions;
+
+    PoolKey(String jdbcUrl, Map<String, ?> dbOptions) {
+        this.jdbcUrl = jdbcUrl;
+        this.dbOptions = dbOptions;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) {
+            return true;
+        }
+        if (!(obj instanceof PoolKey)) {
+            return false;
+        }
+        boolean jdbcUrlEqual = ((PoolKey) obj).jdbcUrl.equals(this.jdbcUrl);
+        return jdbcUrlEqual && dbOptionsEqual((PoolKey) obj);
+    }
+
+    @Override
+    public int hashCode() {
+        int hashCode = 17;
+        hashCode = hashCode * 31 + jdbcUrl.hashCode();
+        if (dbOptions != null) {
+            hashCode = 31 * hashCode + calculateDbOptionsHashCode();
+        }
+        return hashCode;
+    }
+
+    private int calculateDbOptionsHashCode() {
+        int hashCode = 17;
+        for (Map.Entry<String, ?> entry : dbOptions.entrySet()) {
+            int keyHashCode = entry.getKey().hashCode();
+            Object value = entry.getValue();
+            BType type = TypeChecker.getType(value);
+            int typeTag = type.getTag();
+            int valueHashCode;
+            switch (typeTag) {
+            case TypeTags.STRING_TAG:
+            case TypeTags.DECIMAL_TAG:
+                valueHashCode = value.hashCode();
+                break;
+            case TypeTags.BYTE_TAG:
+            case TypeTags.INT_TAG:
+                long longValue = (Long) value;
+                valueHashCode = (int) (longValue ^ (longValue >>> 32));
+                break;
+            case TypeTags.FLOAT_TAG:
+                long longValueConvertedFromDouble = Double.doubleToLongBits((Double) value);
+                valueHashCode = (int) (longValueConvertedFromDouble ^ (longValueConvertedFromDouble >>> 32));
+                break;
+            case TypeTags.BOOLEAN_TAG:
+                valueHashCode = ((Boolean) value ? 1 : 0);
+                break;
+            default:
+                throw new AssertionError("Type " + type.getName() + " shouldn't have occurred");
+            }
+            hashCode = hashCode + keyHashCode + valueHashCode;
+        }
+        return hashCode;
```

Comment:
Can't we simplify this logic by getting the string by `String mapString = dbOptions.entrySet().toString();` and then by getting the hashcode for that string?

Reply:
In that case we need to check whether we'd get the same value for different objects.
If it only depends on values, we should be able to use it.

Reply:
It seems we can directly use the hashCode method of the map itself since it is returning a shallow hashcode[1].

[1] https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#hashCode--

Reply:
Yes. Lets simplify the logic by using it directly.

Reply:
Tried using the hashcode method in the Map itself but test were failing.
In our `MapValueImpl `we've overrided the `hashcode` method[1] which should be the reason for this. Also, `dbOptions.entrySet().toString().hashcode()` fails as well. The `toString` output was different in different cases.

However, if we use `dbOptions.entrySet().hashcode()` there wasn't an issue. Still I think, since we cannot guarantee that there isn't a possibility for this to be overridden, its best if we override it ourselves to suit our needs. This could even break if the `hashcode` method implementation of a Ballerina value changes.

[1] https://github.com/ballerina-platform/ballerina-lang/blob/v0.992.0-m7/bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/MapValueImpl.java#L240